File: 0954-maximum-sum-circular-subarray/0954-maximum-sum-circular-subarray.py
class Solution:
    def maxSubarraySumCircular(self, nums: List[int]) -> int:
        curr=0
        max_sum=float('-inf')

        for i in range(len(nums)):
            if curr+nums[i] < nums[i]:
                curr=nums[i]
            else:
                curr+=nums[i]
            max_sum=max(max_sum,curr)

        total_sum=sum(nums[:])
        curr=0
        min_sum=float('inf')
        for i in range(len(nums)):
            if curr+nums[i] < nums[i]:
                curr+=nums[i]
            else:
                curr=nums[i]
            min_sum=min(min_sum, curr)

        if total_sum==min_sum:
            return max_sum

        return max_sum if max_sum > (total_sum-min_sum) else total_sum-min_sum

File: 0954-maximum-sum-circular-subarray/0954-maximum-sum-circular-subarray.cpp
class Solution {
public:
    int maxSubarraySumCircular(vector<int>& nums) {
        int cmax = 0;
        int cmin = 0;
        int maxSum = nums[0];
        int minSum = nums[0];
        int totalSum = 0;

        for(int num:nums){
            cmax=max(cmax,0) + num;
            maxSum=max(maxSum,cmax);
            
            cmin=min(cmin,0) + num;
            minSum=min(minSum,cmin);

            totalSum+=num;
        }

        if(totalSum==minSum){
            return maxSum;
        }

        return max(maxSum, totalSum-minSum);
    }
};